web3c/contract_factory.js

```diff
@@ -41,7 +67,13 @@ function makeContractFactory(web3, providerFn) {
       provider = providerFn(address, options);
     }
 
+    // setProvider may clear the existing subscriptions, so in order to
+    // keep the state of the provider, we restore them after a call to
+    // setProvider
+    const currentProvider = c.currentProvider;
+    const notificationCallbacks = retrieveNotificationCallbacks(currentProvider);
     c.setProvider.call(this, provider);
+    restoreNotificationCallbacks(currentProvider, notificationCallbacks);
```

Comment:
@armaniferrante please take a look at this restore notifications because I'm not sure it's a good idea, but it's the only one I had to make it work.

The problem is:
1. The provider (a ws provider) has a list of callbacks that are subscribed to events
2. calling `setProvider` resets the list of callbacks to `[]`
3. When an event is received, the `ws` provider has `[]` of callbacks, so the event is not raised. 

By doing this, the `ws` provider has the original list of callbacks available, and can forward the subscription events to the correct request manager.

Reply:
@eauge I'll have more to say after investigating the code, but your description seems reasonable. However, we should at least consider and understand the tradeoffs of a deeper change, namely ensuring that `setProvider` doesn't reset the list of callbacks. 

@willscott is this the same `setProvider` bug you were mentioned today?

Reply:
i was seeing the ws provider actually getting closed as part of the chain when this was triggered, so in a potentially slightly different state.

Reply:
https://github.com/oasislabs/web3c.js/pull/168 will hopefully fix that